### ISSUE TITLE
Make feedback survey optional after heartbeat trigger

### ISSUE THREAD

**Issue by johngruen** [mozilla/testpilot#1327]
Currently we pop a survey straight from the heartbeat rating.
Let's see what happens if we issue a 'thank you' then give users the choice to take a quick survey.

**Comment by sevaan** [comment]
UX: https://mozilla.invisionapp.com/share/Q77S1H6ZC#/187182724_Heartbeat_UI

**Comment by johngruen** [comment]
SHIP IT

**Comment by dannycoates** [comment]
@sevaan what style is that button using? I see we're using the 'heartbeat' style instead of the normal notification bar styles, but I don't see any styles for buttons with [heartbeat](https://github.com/mozilla/gecko-dev/blob/7d334fcb5086d920646ec22f82345bf4eafead42/browser/themes/shared/UITour.inc.css#L157). 

Here's what it looks like as a label with 'text-link' style like the heartbeat uses for 'Learn More'
<img width="425" alt="screen shot 2016-09-09 at 1 40 29 pm" src="https://cloud.githubusercontent.com/assets/87619/18402181/1610ebd8-7693-11e6-88bf-a99f583c0ca2.png">

And a default notification button on Mac:
<img width="438" alt="screen shot 2016-09-09 at 1 57 25 pm" src="https://cloud.githubusercontent.com/assets/87619/18402721/a01fc9d2-7695-11e6-82fb-d8f5abf2f23f.png">

**Comment by sevaan** [comment]
Hey Danny, I just pulled it from my original spec for Heartbeat. They were unused for Heartheat, but I feel like the button styling has been used in other notification bars. @johngruen do you know?

The text link looks OK, but I think if we are going for engagement, a clearer target after the user rates the experiment will help. Can we code it in?

I noticed the Heartbeat UI has been changed slightly since I last worked on it. I've updated the spec and have included some styling details for the button if it helps: https://mozilla.invisionapp.com/share/Q77S1H6ZC#/187620619_Heartbeat_UI_2

**Comment by dannycoates** [comment]
I got this working as two separate prompts because it was easier to implement. Here's an gif of what it looks like in action:

![foo](https://cloud.githubusercontent.com/assets/87619/18445210/5924bb9e-78d1-11e6-8b5e-7944dde1a17e.gif)

@sevaan @johngruen  Is that good enough, or should I spend more effort to get it working in a single notification bar?

**Comment by sevaan** [comment]
I think the two separate prompts is fine. The extra action only makes the change of state even more noticeable.

Could we remove the icon pulse from the Thank You bar?

On Fri, Sep 09, 2016 at 4:45 PM Danny Coates

<
mailto:Danny Coates notifications@github.com

> wrote:

a, pre, code, a:link, body { word-wrap: break-word !important; }

https://github.com/sevaan
what style is that button using? I see we're using the 'heartbeat' style instead of the normal notification bar styles, but I don't see an styles for buttons with
https://github.com/mozilla/gecko-dev/blob/7d334fcb5086d920646ec22f82345bf4eafead42/browser/themes/shared/UITour.inc.css#L157
.

Here's what it looks like as a label with 'text-link' style like the heartbeat uses for 'Learn More'
https://cloud.githubusercontent.com/assets/87619/18402181/1610ebd8-7693-11e6-88bf-a99f583c0ca2.png

—

You are receiving this because you were mentioned.

Reply to this email directly,
https://github.com/mozilla/testpilot/issues/1327#issuecomment-246034829
, or

**Comment by MattGrimes** [comment]
@sevaan @johngruen that is gorgeous. I'm thinking this should be implemented in Heartbeat proper as well. We'll want to keep an eye on what this does to participation rates though. @rayborn do you want to own that?

**Comment by johngruen** [comment]
@MattGrimes, yeah, we had enough people griping about the new tab, that it felt like the right choice. @ckprice maybe we could flag this with a GA event as well?

**Comment by ghost** [comment]
@dannycoates, @johngruen  - While this is a fantastic UX, I have some questions.

Unless I'm misunderstanding the engineering details, the disadvantage here is that **we won't get the HB rating in SurveyGizmo unless the user decides to go to the survey** whereas now we store the rating upon landing.

If we want to keep this superior user flow _and_ don't want to lose ratings then **we should probably consider storing this as a Telemetry event** as we do for the main Firefox Heartbeat.  SurveyGizmo is not intended as an event store so I imagine that we wouldn't want to figure out how to record this Async for non-survey-takers (although there is a write API).

Per the ping for me, I'm happy to do any analysis tasks after this transition as @MattGrimes suggests.

**Comment by dannycoates** [comment]
@rayborn good feedback 👍 

> the engineering details

Now that I've got it working I can explain how it works, then we can adjust it if we want.

The important cases here are:
1. The user rates the experiment and clicks the survey button
2. The user rates the experiment but **doesn't click the survey button**

Case 1 is exactly the same as today. A new tab opens to the survey url passing along some query parameters (including the rating).

Case 2 will cause an [sdk/request](https://developer.mozilla.org/en-US/Add-ons/SDK/High-Level_APIs/request) to make a GET request to the same survey url with the same parameters after the notification gets dismissed or times out. The user won't see the survey, but to SurveyGizmo it should look the same as Case 1. So as you said the rating will get stored on landing.

Now, I made an assumption that I haven't verified yet because I can't see the data on the other end, that SurveyGizmo will treat both cases the same, but all signs in the responses I get suggest they do.

If all this holds true we won't have to make any changes on the reporting side.
